plutus-contract-exe/src/Language/Plutus/Contract/State.hs

```diff
@@ -0,0 +1,324 @@
+[Event]' is that the state may contain arbitrary callbacks introduced by 
+the 'Control.Monad.PromptT' type. The only time we can be certain that a 
+contract has no callbacks that it is waiting for is when 'runContract' (with 'f 
+= Maybe') evaluates to 'Just a'. At that point, the 'a' represents exactly the 
+same state as the list of events, '[Event]', that was used to produce the 'a'. 
+So if we want to serialise the state now, we can just serialise the 'a' and 
+drop all the events.
+
+This is complicated by the fact that contracts may have multiple branches that 
+run at the same time (using '<*>') or one after another (with '>>='). So for a
+contract 'l <*> r', if the 'r' branch is done but the 'l' branch is still 
+waiting for input, we can only drop the events that have been seen by 'r', and 
+we must continue to use the '[Event]' representation for the state of 'l' until 
+both 'l' and 'r' are finished. 
+
+To account for this, we store the events in a 
+'Language.Plutus.Contract.Record.Record' instead of a list. The record is a 
+tree that mirrors the structure of the contract, and whenever one of the 
+contract's branches is done we can serialise its state and store it in the 
+correct location in the record.
+
+Finally we need to know how to serialise the result value of the (branch of a) contract when it's done. This is what the 'StatefulContract' type is for. We can
+use its 'CJSONCheckpoint' constructor to inject the serialisation constraints at
+arbitrary points in the contract. So we could write the previous example as
+'l <*> jsonCheckpoint r' to ensure that the events consumed by the 'r' branch
+are discarded as soon as 'r' produces a result.
+
+To actually run a 'StatefulContract', we need to
+* use 'initialise' to get the initial state (of type 'Record') of the contract
+* for each new event, run 'insertAndUpdate' with the current state and the event
+  to get the new state. 
+
+TODO: There is a lot of boilerplate in the definitions of 'runClosed' and 
+'runOpen' - the cases we have to cover are all possible combinations of the
+constructors of 'StatefulContract' and 'Record'. 
+
+I did attempt to reduce this by using 'newtype Rec i a = Rec Either (OpenRecord i) (ClosedRecord i, a)' and variations but always ran into trouble. I would be 
+very grateful for suggestions on how to improve this situation.
+
+-}
+
+data StatefulContract a where
+    CMap :: (a' -> a) -> StatefulContract  a' -> StatefulContract  a
+    CAp :: StatefulContract  (a' -> a) -> StatefulContract  a' -> StatefulContract  a
+    CBind :: StatefulContract  a' -> (a' -> StatefulContract  a) -> StatefulContract  a
+
+    CContract :: C.ContractPrompt Maybe a -> StatefulContract  a
+    CJSONCheckpoint :: (Aeson.FromJSON a, Aeson.ToJSON a) => StatefulContract  a -> StatefulContract  a
+
+initialise
+    :: ( MonadWriter Hooks m )
+    => StatefulContract a
+    -> m (Either (OpenRecord Event) (ClosedRecord Event, a))
```

Comment:
I agree that this type doesn't seem quite right, but I'm not entirely sure what else should be there.